File: C++/programas/Listas-Pilas dinamicas/Listas-Pilas dinamicas/Pila_Alumnos.cpp
#include <stack> //Cabecera para trabajar con pilas din?micas.
#include <iostream> //Cabecera para entrada y salida de datos.
#include "Pila_Alumnos.h" //Cabecera de la clase Pila_Alumnos.

//Procedimiento que despliega men? de opciones para trabajar con las pilas.
void Pila_Alumnos::Menu(std::stack<double> Prom, std::stack<std::string> Alum) {
    int Op; //Variable que recibe la opci?n escogida por el usuario.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    std::cout << "-Pila de alumnos y sus promedios-\n\n";

    //Despliega el men? de opciones.
    printf("Nombres registrados: %d.\n", Alum.size());
    std::cout << "\n-Men?-\n1) Insertar alumno.\n2) Eliminar alumno.\n3) Vaciar pila.\n4) Buscar alumno.\n5) Recorrer pila.\n6) Salir.\n";
    //Ingreso de valor a variable.
    std::cout << "Opci?n: "; std::cin >> Op;

    if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
    {
        std::cout << "Error de formato. ";
        system("pause"); //Pausa el flujo del programa.
        std::cin.clear(); //Limpia el buffer de cin.
        //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        Menu(Prom, Alum); //Llamada recursiva a Menu.
    }
    else
    {
        //Selectiva switch para saltar al m?todo escogido por el usuario.
        switch (Op) {
            case 1:
                AgregarAlumnos(Prom, Alum);
                break;
            case 2:
                EliminarAlumno(Prom, Alum);
                break;
            case 3:
                VaciarPila(Prom, Alum);
                break;
            case 4:
                BuscarAlumno(Prom, Alum);
                break;
            case 5:
                RecorrerAlumno(Prom, Alum);
                break;
            case 6:
                exit(-1); //Salida del programa
                break;
            default:
                std::cout << "Selecciona una opci?n del men? mostrado.";
                system("pause"); //Pausa el flujo del programa.
                Menu(Prom, Alum); //Llamada recursiva a Menu.
        }
    }
}

//Procedimiento que ingresa valores al tope de la pila.
void Pila_Alumnos::AgregarAlumnos(std::stack<double> Prom, std::stack<std::string> Alum) {
    std::string nombre; //Variable que recibe nombre para la pila string.
    double promedio; //Variable que recibe un promedio para la pila double.
    int Op; //Variable que recibe la opci?n escogida por el usuario.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    std::cout << "--Inserci?n de alumnos con sus promedios--\n\n";
    //Ingreso de valores a variables.
    std::cout << "Ingresa un nombre: "; std::cin >> nombre;
    std::cout << "Ingresa un promedio: "; std::cin >> promedio;

    if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
    {
        std::cout << "Error de formato. ";
        system("pause"); //Pausa el flujo del programa.
        std::cin.clear(); //Limpia el buffer de cin.
        //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        Menu(Prom, Alum); //Llamada a Menu.
    }
    else
    {
        Alum.push(nombre); //Agrega elemento a la pila.
        Prom.push(promedio); //Agrega elemento a la pila.

        //Ingreso de valor a variable.
        std::cout << "\nIngresado. ?Desea seguir agregando alumnos? Si[1]: "; std::cin >> Op;
        if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
        {
            std::cout << "Error de formato. ";
            system("pause"); //Pausa el flujo del programa.
            std::cin.clear(); //Limpia el buffer de cin.
            //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            Menu(Prom, Alum); //Llamada a Menu.
        }
        else
        {
            //Seg?n la opci?n escogida, se salta a un m?todo de la clase.
            if (Op == 1)
                AgregarAlumnos(Prom, Alum);
            else
                Menu(Prom, Alum);
        }
    }
}

//Procedimiento que elimina el elemento tope de la pila.
void Pila_Alumnos::EliminarAlumno(std::stack<double> Prom, std::stack<std::string> Alum) {
    int Op; //Variable que recibe la opci?n escogida por el usuario.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la pila est? vac?a...
    if (Alum.empty()) {
        std::cout << "Pila vac?a. ";
        system("pause"); //Pausa el flujo del programa.
        Menu(Prom, Alum); //Llamada a Menu.
    }
    else {
        std::cout << "--Eliminar un alumno--\n\n";
        //Despliega mensaje con los elementos a eliminar.
        std::cout << "El alumno " << Alum.top() << " con promedio " << Prom.top() << " fue eliminado. ";
        system("pause"); //Pausa el flujo del programa.
        Alum.pop(); //Elimina el ?ltimo elemento ingresado.
        Prom.pop(); //Elimina el ?ltimo elemento ingresado.
        //Ingreso de valor a variable.
        std::cout << "?Desea eliminar otro ? Si[1] : "; std::cin >> Op;
        if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
        {
            std::cout << "Error de formato. ";
            system("pause"); //Pausa el flujo del programa.
            std::cin.clear(); //Limpia el buffer de cin.
            //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            Menu(Prom, Alum); //Llamada a Menu.
        }
        else
        {
            //Seg?n la opci?n escogida, se salta a un m?todo de la clase.
            if (Op == 1)
                EliminarAlumno(Prom, Alum);
            else
                Menu(Prom, Alum);
        }
    }
}

//Procedimiento que vac?a los elementos de las pilas.
void Pila_Alumnos::VaciarPila(std::stack<double> Prom, std::stack<std::string> Alum) {
    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la pila est? vac?a...
    if (Alum.empty())
        std::cout << "Pila vac?a. ";
    else {
        //Mientras la pila no est? vac?a, se van eliminando los elementos tope de la pila.
        while (!Alum.empty()) {
            Alum.pop(); //Elimina el ?ltimo elemento ingresado.
            Prom.pop(); //Elimina el ?ltimo elemento ingresado.
        }
        std::cout << "Pila vaciada. ";
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Prom, Alum); //Llamada a Menu.
}

//Procedimiento que busca el nombre de un alumno en la pila.
void Pila_Alumnos::BuscarAlumno(std::stack<double> Prom, std::stack<std::string> Alum) {
    std::stack<double> t_prom = Prom;
    std::stack<std::string> t_alum = Alum; //Pilas din?micas adicionales para buscar.
    std::string alum; //Variable auxiliar para buscar el nombre.
    double prom; //Variable auxiliar que recibe el promedio del nombre encontrado.
    bool encontrado = false; //Variable auxiliar para la b?squeda del nombre.
    int Op; //Variable que recibe la opci?n escogida por el usuario.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la pila est? vac?a...
    if (t_alum.empty()) {
        std::cout << "Pila vac?a. ";
        system("pause"); //Pausa el flujo del programa.
        Menu(Prom, Alum); //Llamada a Menu.
    }
    else {
        std::cout << "--B?squeda de alumno--\n\n";
        //Ingreso de valor a variable.
        std::cout << "Ingresa el alumno a buscar: "; std::cin >> alum;
        //Mientras la pila no est? vac?a...
        while (!t_alum.empty()) {
            //Busca al alumno en la pila trabajando ambas pilas de alumnos y promedios.
            //Si encuentra al alumno, almacena tambi?n el promedio de una variable adicional.
            if (t_alum.top().compare(alum) == 0) {
                encontrado = true;
                prom = t_prom.top();
            }
            t_alum.pop(); //Elimina el ?ltimo elemento ingresado.
            t_prom.pop(); //Elimina el ?ltimo elemento ingresado.
        }

        //Si encontr? un nombre, lo despliega
        if (encontrado)
            std::cout << "Se encontr? al alumno. Tiene promedio de " << prom << ". ";
        else
            std::cout << "No se encontr? el alumno. ";

        std::cout << "?Desea buscar otro? Si[1]: "; std::cin >> Op;
        if (std::cin.fail()) //Si se ingres? un tipo de dato incorrecto en la variable.
        {
            std::cout << "Error de formato. ";
            system("pause"); //Pausa el flujo del programa.
            std::cin.clear(); //Limpia el buffer de cin.
            //Ignora el resto de caracteres ingresados en cin hasta el pr?ximo salto de l?nea.
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            Menu(Prom, Alum); //Llamada a Menu.
        }
        else
        {
            //Seg?n la opci?n escogida, se salta a un m?todo de la clase.
            if (Op == 1)
                BuscarAlumno(Prom, Alum);
            else
                Menu(Prom, Alum);
        }

    }
}

//Procedimiento que recorre y despliega los elementos de la pila.
void Pila_Alumnos::RecorrerAlumno(std::stack<double> Prom, std::stack<std::string> Alum) {
    std::stack<double> t_prom = Prom;
    std::stack<std::string> t_alum = Alum; //Pilas din?micas adicionales para buscar.
    int i = 1; //Variable para las iteraciones del ciclo.

    system("cls"); //Limpia la terminal.
    std::locale::global(std::locale("spanish")); //Instrucci?n para utilizar caracteres en espa?ol.

    //Si la pila est? vac?a, despliega mensaje.
    if (t_alum.empty())
        std::cout << "Pila vac?a. ";
    else {
        std::cout << "--Pilas de alumnos con sus promedios--\n\n";
        //Mientras el ciclo no est? vac?o, despliega los elementos de las pilas.
        while (!t_alum.empty()) {
            std::cout << i << " - " << t_alum.top() << " con promedio " << t_prom.top() << ".\n";
            t_alum.pop();
            t_prom.pop();
            i++; //Incrementa contador.
        }

        std::cout << "\n";
    }

    system("pause"); //Pausa el flujo del programa.
    Menu(Prom, Alum); //Llamada a Menu.
}

File: Java/programas/Pila dinámica/src/Pila_Alumnos.java
import java.io.IOException; //Excepción si hay error durante entrada y salida de datos.
import java.util.InputMismatchException; //Excepción si hay un error de formato en tipos de datos.
import java.util.Stack; //Clase para trabajar con pilas dinámicas.
import java.util.Scanner; //Librería para declara objeto Scanner de entrada y salida.

public class Pila_Alumnos {
    public Scanner lectura = new Scanner(System.in); //Declara objeto Scanner.
    
    //Procedimiento que despliega un menú con distintas opciones para realizar con una pila dinámica.
    public void Menu(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        int Op;

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        System.out.println("-Pila de alumnos con sus promedios-\n");
        //Selectiva try-catch por si se ingresa un tipo de dato incorrecto en la variable.
        try{
            System.out.printf("Alumnos registrados: %s\n", Alum.size());
            System.out.println("\n-Menú-\n1) Insertar alumno.\n2) Eliminar alumno.\n3) Vaciar pila.\n4) Buscar alumno.\n5) Recorrer pila.\n6) Salir.");
            System.out.print("Selecciona una opción: "); Op = lectura.nextInt();
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            //Selectiva switch para determinar qué método será llamado a partir del menú.
            switch(Op){
                case 1:
                    AgregarAlumno(Prom, Alum); //Llamada al procedimiento que agrega elementos a las pilas Prom y Alum.
                    break;
                case 2:
                    EliminarAlumno(Prom, Alum); //Llamada al procedimiento que elimina elementos a las pilas Prom y Alum.
                    break;
                case 3:
                    VaciarPila(Prom, Alum); //Llamada al procedimiento que vacía las pilas Prom y Alum.
                    break;
                case 4:
                    BuscarAlumno(Prom, Alum); //Llamada al procedimiento que busca elementos a las pilas Prom y Alum.
                    break;
                case 5:
                    RecorrerAlumnos(Prom, Alum); //Llamada al procedimiento que recorre los elementos a las pilas Prom y Alum.
                    break;
                case 6:
                    lectura.close(); //Cierra el objeto Scanner.
                    System.exit(0); //Cierra el programa.
                    break;
                default:
                    System.out.println("Selecciona una opción del menú mostrado.");
                    lectura.nextLine(); //Limpia buffer de entrada de datos.
                    Menu(Prom, Alum); //Llamada recursiva a Menu.
            }
        } catch(InputMismatchException e){
            System.out.println("Error de formato.");
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            Menu(Prom, Alum); //Llamada recursiva a Menu.
        }
    }

    //Procedimiento que agrega un elemento a una pila dinámica.
    public void AgregarAlumno(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        int Op;
        String nombre;
        double promedio;

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.
        //Selectiva try-catch por si se ingresa un tipo de dato incorrecto en la variable.
        try{
            System.out.print("Ingresa un nombre: "); nombre = lectura.nextLine();
            System.out.print("Ingresa un promedio: "); promedio = lectura.nextDouble();
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            
            Alum.push(nombre); //Agrega elemento String a la pila.
            Prom.addElement(promedio); //Agrega un objeto a la pila.
            System.out.print("\nIngresado. ¿Desea seguir agregando alumnos? Si[1]: "); Op = lectura.nextInt();
            lectura.nextLine(); //Limpia buffer de entrada de datos.

            //Si se decidió agregar otro elemento.
            if (Op == 1)
                AgregarAlumno(Prom, Alum); //Llamada recursiva a AgregarAlumno.
            else
                Menu(Prom, Alum); //Llamada a Menu.
        } catch(InputMismatchException e){
            System.out.println("Error de formato.");
            lectura.nextLine(); //Limpia buffer de entrada de datos.
            AgregarAlumno(Prom, Alum); //Llamada recursiva a AgregarAlumno.
        }
    }

    //Procedimiento que elimina un elemento a una pila dinámica.
    public void EliminarAlumno(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si el tamaño de la pila es 0, despliega mensaje que está vacía.
        if (Alum.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        } else{
            //Despliega mensaje de qué elementos de la pila serán borrados (tope).
            System.out.printf("El alumno %s con promedio %s fue elimniado. Presiona ENTER para regresar al menú...", Alum.peek(), Prom.peek());
            Alum.pop(); //Elimina el objeto en la cima de la pila.
            Prom.pop(); //Elimina el objeto en la cima de la pila.
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        }
    }

    //Procedimiento que borra todos los elementos de una pila dinámica.
    public void VaciarPila(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si el tamaño de la pila es 0, despliega mensaje que está vacía.
        if (Alum.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        } else{
            Alum.clear(); //Limpia la pila.
            Prom.clear(); //Limpia la pila.
            //Despliega mensaje de que la pila fue vaciada.
            System.out.print("Las pilas Alumnos y Promedios fueron vaciadas. Presiona ENTER para regresar al menú...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        }
    }

    //Procedimiento que busca un elemento de una pila dinámica.
    public void BuscarAlumno(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        String nombre;

        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si el tamaño de la pila es 0, despliega mensaje que está vacía.
        if (Alum.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        } else{
            System.out.print("Ingresa el nombre a buscar: "); nombre = lectura.nextLine(); //Recibe nombre a buscar en la pila.
            if (Alum.contains(nombre) == true){ //Busca el nombre en la pila, regresa true si lo encontró.
                //Despliega un mensaje diciendo si encontró el nombre y en qué posición lo halló.
                System.out.printf("Se encontró el alumno, tiene promedio de %s.", Prom.elementAt(Alum.indexOf(nombre)));
                System.out.print("\nPresione ENTER para regresar al menú...");
                System.in.read(); //Espera a que se presione ENTER para continuar.
                Menu(Prom, Alum); //Llamada a Menu.
            } else
                System.out.print("No se encontró el alumno. Presione ENTER para regresar al menú...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        }
    }

    //Procedimiento que recorre los elementos de una pila dinámica.
    public void RecorrerAlumnos(Stack<Double> Prom, Stack<String> Alum) throws InterruptedException, IOException{
        int i;
        
        new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor(); //Limpia consola/terminal/pantalla.

        //Si el tamaño de la pila es 0, despliega mensaje que está vacía.
        if (Alum.size() == 0){
            System.out.print("Pila vacia. Presione ENTER para continuar...");
            System.in.read(); //Espera a que se presione ENTER para continuar.
            Menu(Prom, Alum); //Llamada a Menu.
        } else{
            System.out.println("---Alumnos con sus promedios---\n");
            for(i = (Alum.size() - 1); i >= 0; i--){ //Despliega todos los elementos de las dos pilas.
                System.out.printf("%s - %s\n", Alum.elementAt(i), Prom.elementAt(i));
            }
        }
        System.out.print("\nPresione ENTER para regresar al menú...");
        System.in.read(); //Espera a que se presione ENTER para continuar.
        Menu(Prom, Alum); //Llamada a Menu.
    }
}